File: DSA/GEHNA_KHANNA_590013140/day38question2.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 1000

int mainStack[MAX], minStack[MAX];
int topMain = -1, topMin = -1;

void push(int x) {
    mainStack[++topMain] = x;

    
    if (topMin == -1 || x <= minStack[topMin]) {
        minStack[++topMin] = x;
    }
}

void pop() {
    if (topMain == -1) {
        printf("Stack is empty\n");
        return;
    }

    int popped = mainStack[topMain--];

    if (popped == minStack[topMin]) {
        topMin--;  // Pop from min stack too
    }
}

int top() {
    if (topMain == -1) {
        printf("Stack is empty\n");
        return -1;
    }
    return mainStack[topMain];
}

int getMin() {
    if (topMin == -1) {
        printf("Stack is empty\n");
        return -1;
    }
    return minStack[topMin];
}

int main() {
    int choice, val;

    while (1) {
        printf("\n1. Push\n2. Pop\n3. Top\n4. Get Min\n5. Exit\nEnter choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter value to push: ");
                scanf("%d", &val);
                push(val);
                break;
            case 2:
                pop();
                break;
            case 3:
                printf("Top element: %d\n", top());
                break;
            case 4:
                printf("Minimum element: %d\n", getMin());
                break;
            case 5:
                exit(0);
            default:
                printf("Invalid choice!\n");
        }
    }

    return 0;
}

File: DSA/N_PARTH_PANDEY_590012761/Day38Question2.py
class MinStack:
    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, val):
        self.stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1][0]:
            # Store both the value and its count
            if self.min_stack and val == self.min_stack[-1][0]:
                self.min_stack[-1] = (val, self.min_stack[-1][1] + 1)
            else:
                self.min_stack.append((val, 1))

    def pop(self):
        if not self.stack:
            return None
        val = self.stack.pop()
        if val == self.min_stack[-1][0]:
            if self.min_stack[-1][1] > 1:
                self.min_stack[-1] = (val, self.min_stack[-1][1] - 1)
            else:
                self.min_stack.pop()
        return val

    def top(self):
        if not self.stack:
            return None
        return self.stack[-1]

    def getMin(self):
        if not self.min_stack:
            return None
        return self.min_stack[-1][0]


if __name__ == "__main__":
    min_stack = MinStack()
    operations = []
    
    print("Enter stack operations (push <nums>, pop, top, getMin, or 'exit' to quit):")
    while True:
        operation = input("> ").strip().lower()
        if operation == 'exit':
            break
        
        if operation.startswith('push'):
            try:
                nums = list(map(int, operation.split()[1:]))
                for num in nums:
                    min_stack.push(num)
                    operations.append(f"push {num}")
            except (IndexError, ValueError):
                print("Invalid push operation. Usage: push <num1> <num2> ...")
        elif operation == 'pop':
            result = min_stack.pop()
            operations.append("pop")
            print(f"Popped: {result}")
        elif operation == 'top':
            result = min_stack.top()
            operations.append("top")
            print(f"Top element: {result}")
        elif operation == 'getmin':
            result = min_stack.getMin()
            operations.append("getMin")
            print(f"Minimum element: {result}")
        else:
            print("Invalid operation. Use push, pop, top, getMin, or exit")
    
    print("\nOperation history:")
    for op in operations:
        print(op)